### PR TITLE
accelerate deepspeed and gradient accumulation integrate

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Shift deepspeed integration to accelerate
2. Shift Gradient Accumulation to Accelerate
3. Merge after #23168 
4. no user facing change. Now user can use `accelerate launch` with trainer for DeepSpeed, e.g.:
```
accelerate launch --num_processes=2 --mixed_precision=bf16 --use_deepspeed --gradient_accumulation_steps=1 --gradient_clipping=1 --zero3_init_flag=True --zero3_save_16bit_model=False --zero_stage=3 --offload_optimizer_device=none --offload_param_device=none ./examples/pytorch/text-classification/run_glue.py  --model_name_or_path bert-base-cased   --task_name $TASK_NAME   --do_train   --do_eval   --max_seq_length 128   --per_device_train_batch_size 16   --learning_rate 5e-5   --num_train_epochs 3   --output_dir /tmp/$TASK_NAME/ --overwrite_output_dir --bf16
```

Usual run using `torchrun` and trainer args is unimpacted:
```
torchrun --nnodes 1 --nproc-per-node 2 ./examples/pytorch/text-classification/run_glue.py  --model_name_or_path bert-base-cased   --task_name $TASK_NAME   --do_train   --do_eval   --max_seq_length 128   --per_device_train_batch_size 16   --learning_rate 5e-5   --num_train_epochs 3   --output_dir /tmp/$TASK_NAME/ --overwrite_output_dir --deepspeed ~/transformers/tests/deepspeed/ds_config_zero2.json
```
5. Save and load utils are changed accordingly

### CODE REVIEW

**Review by muellerzr — APPROVED** [review]
Thanks for working on this! Looks great, once tests pass fully :)

File: src/transformers/trainer.py
Comment (muellerzr) [inline src/transformers/trainer.py:2708]: Accelerate already does this when we have enabled mixed precision during `backward`, so this probably isn't behaving properly/applying twice, no?

**Review by muellerzr — APPROVED** [review]
LG2M, one clarification which indeed is true

**Review by sgugger — APPROVED** [review]
Thanks for working on this. Is the diff longer than expected because of other PRs to be merged before?

Might be cool to Have Stas have a look (not pinging him here too early) once this is ready to merge and tests are confirmed to all pass.

**Comment by pacman100** [comment]
> Thanks for working on this. Is the diff longer than expected because of other PRs to be merged before?

Due to updating from main, it is not showing the diff wrt previous branches. Weird. 

> Might be cool to Have Stas have a look (not pinging him here too early) once this is ready to merge and tests are confirmed to all pass.

Yes, definitely. All tests are passing already. Checked the slow tests offline.

**Comment by pacman100** [comment]
@sgugger, now the diff is only specific to DeepSpeed changes + gradient accumulation changes + saving/loading changes wrt previous PR.

**Comment by pacman100** [comment]
Hello @stas00, please review this PR which aims to shift the accelerate handling in Trainer to Accelerate. Thank you!

File: docs/source/en/main_classes/deepspeed.mdx
Comment (stas00) [inline docs/source/en/main_classes/deepspeed.mdx:1703]: Does this mean that this change will break BC? 

I thought breaking BC was definitely not allowed.

File: src/transformers/deepspeed.py
Comment (stas00) [inline src/transformers/deepspeed.py:271]: this would break BC

**Review by stas00 — COMMENTED** [review]
@pacman100, I think the main concern is that this PR appears to be breaking BC in at least a few places - please correct me if I'm wrong. I think those cases are super minor and probably won't cause too many breakages for the users if any. I'll leave up to you to decide.

As I don't know all the nuances of Accelerate's Deepspeed integration I can't do a detailed review, but if all the SLOW tests pass it should be good.

request: as Accelerate is taking over please remove me from the Issues/PR templates as deepspeed integration maintainer while you're at it, since I won't be able to support users any longer.

Thank you!

File: docs/source/en/main_classes/deepspeed.mdx
Comment (stas00) [reply docs/source/en/main_classes/deepspeed.mdx:1703]: but this one is probably super-minor - I can't guess but my hunch is that probably barely anybody is using this code.

File: src/transformers/deepspeed.py
Comment (pacman100) [reply src/transformers/deepspeed.py:271]: Yes, this is difficult to fix because there isn't a way to pass a callable to `accelerator.prepare` and know the purpose of the callable

**Review by muellerzr — APPROVED** [review]
Thanks for doing this! Looks great

**Pull request merged**
